### ISSUE TITLE
Implementation of Gosling with Codebase Change

### ISSUE THREAD

**Issue by Sammysupport** [blueprint-freespeech/ricochet-refresh#192]
Hello,

after studying Gosling a bit, the idea came up to switch the code base for RicoChet Refresh:

like Freenet has done: 
https://www.mail-archive.com/devl@freenetproject.org/msg55262.html 

As new codebase for RicoChet, the very advanced elaborated Messenger Spot-On could be used:
https://github.com/textbrowser/spot-on
https://www.reddit.com/r/Spot_On_Encryption/

Request: 
- Please add Gosling to this base of code.
- Rename the result to "Go Spot-On RicoChet", if not a community choosen name suggestion and then name election is created ( also the name "R2E2" (working title) could be an suggested alternative..).

To develop the old R2 messengers Gui codebase further, would need probably more work rather than to uncomment and blind out some (unwished or too detailed) gui elements of Spot-On you want to hide in the compile process.

Thanks for a consideration and wise decision, which of course can be only made, if there has been an extensive exploration of that  messaging system:

Can you provide a public Echo server reachable by Tor to test that also with R2 community members?
When is the audit of Gosling ready to "Go-Spot-On-RicoChet"?

**Comment by Sammysupport** [comment]
just wanted to add, that both pathes for Gosling with Codebasechange are possible:

Default: Codebase Client - Gosling - Tor - Tor - Gosling - Codebase Client
Optional: Codebase Client - Gosling - Tor - Echo-Server - Tor - Gosling - Codebase Client.

that would be the best way for all aspects. 

(#190 could then be probably the preferred way - as well in a codebase change - to be evaluated (or not and just estimated by personal insights) with results for better latency and reliability.)

**Comment by pospeselr** [comment]
I"m failing to understand what the point of adding this other thing to the mix is. What security/privacy properties are added here?

**Comment by Sammysupport** [comment]
we should run a test with that client and code basis. there are much features:

- multi encryption: you send a three times encrypted packet over a gosling to gosling tunnel over Tor
- first desktop client with McEliece Algroithm (currently linux, windows in an older version)
- Cryptographic Calling, means you renew the encrypted channel and can switch from asym to sym, can switch from longterm key to temp keys
- SMP Socialist Millionaire Calling, create a channel without transmitting a password
- Zero Knowledge proofs of Chat partner
- File Transfer in the Chat window
- Opening a Group chat out of the 1:1 Friends Chat window
- if running (network connected), no need for a central or any server for groupchat
- if running (network connected), no need for a central or any server for messaging to offline friends
- if the codebasis connects over gosling to gosling, you need not any server or any IP to connect for chat, means also for groupchat or offlinechat
- Chat is extended with p2p email, so it is messaging in one term
- Secret Stream keys, many temp keys per one message, not one key per message or session
- individual set of algorithms and modern hash algos
- Strong RSA possible
- Human Proxies for inner envelopes: send a message not from your node, but from a freinds node
- encryption of chat data at rest on the harddisk
- Gui Lock
- because of the echo protocol used, there is no metadata
- this means also you do not know which user is reading which packet
- there are no IP addresses in the encrypted http packets needed, hence it is beyond cryptographic routing.
- just to mention a few..

in summary there is a lot of security and privacy you get with a codebasechange and implementing Gosling there for a R3 client.

**Comment by Sammysupport** [comment]
1:1 chat
![image](https://github.com/user-attachments/assets/52fd9fed-301b-4a6c-8b82-9cd0e6b12c03)


groupchat
![image](https://github.com/Sammysupport/spot-on/assets/153425149/e6cf48bf-ff61-4516-9836-f0611b33ff2d)
